internal/database/testing/database.go

```diff
@@ -51,3 +52,60 @@ func DumpTable(c *gc.C, db *sql.DB, table string) {
 	writer.Flush()
 	fmt.Fprintln(os.Stdout)
 }
+
+// DumpTablesJSON prints the named tables to stdout in a JSON format. This is
+// intended as a debugging tool for state tests. It is not intended for use in
+// production code.
+func DumpTablesJSON(c *gc.C, db *sql.DB, tableNames ...string) {
+	for _, tableName := range tableNames {
+		query := fmt.Sprintf("SELECT * FROM %s", tableName)
+		table := &tableData{
+			tableName: tableName,
+		}
+
+		rows, err := db.Query(query)
+		c.Assert(err, jc.ErrorIsNil)
+
+		columns, err := rows.Columns()
+		c.Assert(err, jc.ErrorIsNil)
+		table.columnNames = columns
+
+		for rows.Next() {
+			row := make([]any, len(columns))
+			for i := range row {
+				row[i] = new([]byte)
+			}
+			err = rows.Scan(row...)
+			c.Assert(err, jc.ErrorIsNil)
+			table.rows = append(table.rows, row)
+		}
+
+		printJSON(c, table)
+	}
+
+}
+
+// tableData is an intermediate representation of an arbitrary table, ready to
+// be processed for printing.
+type tableData struct {
+	tableName   string
+	columnNames []string
+	rows        [][]any
+}
+
+func printJSON(c *gc.C, table *tableData) {
+	fmt.Printf("***** TABLE %q *****\n", table.tableName)
```

Comment:
You are not really printing JSON when this line is in here. I think if we are going to say this is doing json make it all json.

Turn the struct into an object of table name keys that are arrays of objects. That way it is all valid from top to bottom.

Reply:
This is solely for debugging. No one should be treating the output of this function as valid JSON, and I don't want to encourage that either. The fact that it resembles JSON is coincidental. I could have gone with any format that displays the data in this way.

I don't think your proposal would achieve much other than make the output less readable. If your problem is calling the function "JSON" when it's not actually valid JSON, then sure, I'm open to other name suggestions.